tests/pamiq_core/torch/test_model.py

```diff
@@ -0,0 +1,199 @@
+        assert not output_tensor.requires_grad
+        assert output_tensor.grad_fn is None
+        # check if backward results in an error.
+        with pytest.raises(RuntimeError):
+            output_tensor.backward()
+
+
+class TestTorchTrainingModel:
+    @pytest.fixture
+    def model(self) -> nn.Module:
+        return nn.Linear(3, 5)
+
+    @pytest.fixture
+    def torch_training_model_default(self, model: nn.Module) -> TorchTrainingModel:
+        return TorchTrainingModel(
+            model, has_inference_model=True, inference_thread_only=False
+        )
+
+    @pytest.fixture
+    def torch_training_model_with_no_inference_model(
+        self, model: nn.Module
+    ) -> TorchTrainingModel:
+        return TorchTrainingModel(
+            model, has_inference_model=False, inference_thread_only=False
+        )
+
+    @pytest.fixture
+    def torch_training_model_inference_only(
+        self, model: nn.Module
+    ) -> TorchTrainingModel:
+        return TorchTrainingModel(
+            model, has_inference_model=True, inference_thread_only=True
+        )
+
+    @pytest.fixture
+    def torch_training_models(
+        self,
+        torch_training_model_default: TorchTrainingModel,
+        torch_training_model_with_no_inference_model: TorchTrainingModel,
+        torch_training_model_inference_only: TorchTrainingModel,
+    ) -> list[TorchTrainingModel]:
+        return [
+            torch_training_model_default,
+            torch_training_model_with_no_inference_model,
+            torch_training_model_inference_only,
+        ]
+
+    def test_create_inference(
+        self, torch_training_models: list[TorchTrainingModel]
+    ) -> None:
+        for torch_training_model in torch_training_models:
+            if torch_training_model.has_inference_model:
+                torch_inference_model = torch_training_model.inference_model
+                # check if the internal models are same
+                assert torch.equal(
+                    torch_training_model.model.weight,
+                    torch_inference_model._raw_model.weight,
+                )
+                # check about pointers
+                if torch_training_model.inference_thread_only:
+                    assert (
+                        torch_training_model.model is torch_inference_model._raw_model
+                    )
+                else:
+                    assert (
+                        torch_training_model.model
+                        is not torch_inference_model._raw_model
+                    )
+
+    def test_forward(
+        self, model: nn.Module, torch_training_models: list[TorchTrainingModel]
+    ) -> None:
+        for torch_training_model in torch_training_models:
+            input_tensor = torch.randn([2, 3])
+            output_tensor = torch_training_model.forward(input_tensor)
+            expected_tensor = model(input_tensor)
+            assert torch.equal(output_tensor, expected_tensor)
```

Comment:
下手にforを使わない。ロジックを複雑化させない。これはdefault modelに関してのみテストすれば十分。
```suggestion
    def test_forward(
        self, model: nn.Module, torch_training_model_default
    ) -> None:
        input_tensor = torch.randn([2, 3])
        output_tensor = torch_training_model.forward(input_tensor)
        expected_tensor = model(input_tensor)
        assert torch.equal(output_tensor, expected_tensor)
```

Reply:
default model以外の設定でも`forward`が動くかどうかって確認した方がいい気がする　例えば`has_inference_model=False`にしたら`forward`が意図通り動きませんでしたみたいなのが起きても困るし
あと関数の頭の、テスト対象を選択するforくらいは許容してほしさあるかも　もしこれがないと全く同じテストを3種類のモデルについてベタ書きする訳だけど、for使う場合よりかえって見通し悪くならない？

Reply:
OK